stdlib/universe/join_test.flux

```diff
@@ -41,21 +42,59 @@ outData =
 ,,1,RAM,2018-05-22T19:53:46Z,3,4,user1,user2,b
 ,,1,RAM,2018-05-22T19:53:56Z,5,0,user1,user2,b
 "
-t_join = (table=<-) => {
+
+testcase join_base {
+    input = testing.loadStorage(csv: inData)
+    want = testing.loadMem(csv: outData)
     left =
-        table
+        input
             |> range(start: 2018-05-22T19:53:00Z, stop: 2018-05-22T19:55:00Z)
             |> drop(columns: ["_start", "_stop"])
             |> filter(fn: (r) => r.user == "user1")
             |> group(columns: ["user"])
     right =
-        table
+        input
             |> range(start: 2018-05-22T19:53:00Z, stop: 2018-05-22T19:55:00Z)
             |> drop(columns: ["_start", "_stop"])
             |> filter(fn: (r) => r.user == "user2")
             |> group(columns: ["_measurement", "_field"])
 
-    return join(tables: {left: left, right: right}, on: ["_time", "_measurement", "_field"])
+    got = join(tables: {left: left, right: right}, on: ["_time", "_measurement", "_field"])
+
+    testing.diff(want: want, got: got)
 }
 
-test _join = () => ({input: testing.loadStorage(csv: inData), want: testing.loadMem(csv: outData), fn: t_join})
+testcase join_repro_4692 {
+    // Running `findRecord` repeatedly on the same stream can clone the
+    // inputs to a join, mutating the original spec each time to refer to more
+    // and more input streams each time. This leads to a panic.
+    //
+    // This tests verifies that repeated runs of `findRecord` on a `join` is
+    // "safe" to do in terms of "it no longer panics."
+    // Refs: <https://github.com/influxdata/flux/issues/4692>
+    xs = array.from(rows: [{id: 1, x: 1}, {id: 2, x: 2}, {id: 3, x: 3}])
+    ys = array.from(rows: [{id: 1, y: 1}, {id: 2, y: 2}, {id: 3, y: 3}])
+    zs = join(tables: {xs: xs, ys: ys}, on: ["id"])
+
+    getById = (id) => {
+        r = zs |> filter(fn: (r) => r.id == id) |> findRecord(fn: (key) => true, idx: 0)
+
+        return {x: r.x, y: r.y}
+    }
+
+    got =
+        array.from(
+            rows: [
+                {_value: 1},
+                {_value: 2},
+                // repeated lookups for id=2 should be okay...
+                {_value: 2},
+                {_value: 3},
+            ],
+        )
+            |> map(fn: (r) => getById(id: r._value))
+
+    want = array.from(rows: [{x: 1, y: 1}, {x: 2, y: 2}, {x: 2, y: 2}, {x: 3, y: 3}])
+
+    testing.diff(want: want, got: got)
```

Comment:
Can you add another test case and we can see how it goes?
```js
testcase join_repro_4692 {
    xs = array.from(rows: [{id: 1, x: 1, v: "hi"}, {id: 2, x: 2, v: "hi"}, {id: 3, x: 3, v: "hi"}])
    ys = array.from(rows: [{id: 1, y: 1}, {id: 2, y: 2}, {id: 3, y: 3}])

    getById = (id) => {
        zs = if id == 2 then
            join(tables: {xs: xs |> map(fn: (r) => ({r with v: "hey"})), ys: ys}, on: ["id"])
        else
            join(tables: {xs: xs, ys: ys}, on: ["id"])
            
        r = zs |> filter(fn: (r) => r.id == id) |> findRecord(fn: (key) => true, idx: 0)
        return {x: r.x, y: r.y, v: r.v}
    }

    got =
        array.from(
            rows: [
                {_value: 1},
                {_value: 2},
                // repeated lookups for id=2 should be okay...
                {_value: 2},
                {_value: 3},
            ],
        )
            |> map(fn: (r) => getById(id: r._value))

    want = array.from(rows: [{x: 1, y: 1, v: "hi"}, {x: 2, y: 2, v: "hey"}, {x: 2, y: 2, v: "hey"}, {x: 3, y: 3, v: "hi"}])

    testing.diff(want: want, got: got)
```
I am sure that there are typos in this example but the basic idea is that the function that returns a result from `findRecord` will produce a graph whose shape depends on the parameter `id` at runtime. That is to say, the input to the join when `id == 2` will make include `map`, and you can tell if `map` was correctly executed when `v` changes from `hi` to `hey`.

My suspicion is that you're right about needing to do a deep copy, and this new test case would fail as a result.

Reply:
The branching `getById` version is currently passing...
I wondered if this is only because (as written) the joins are defined within the function call, and so are 1:1 with the `findRecord` calls.

I tried moving the joins out to the script root and used the branching to alternate between those variables instead, but that also passed.

Even if these specific cases are passing, it does seem like a deep copy of the spec is desired -- if for no other reason than to justify the copies of the table inputs. I'm just not sure how to achieve this based on the flow we have today.

Reply:
What happens if you do the join inside of the call to map?

Reply:
e01529b defines the joins inside the map fn (assuming I understood you correctly). Still passes. :shrug: